### DISCUSSION TITLE
#96 border-radius bug another example

### DISCUSSION THREAD

**Discussion by rafehimself** [cssbattle/forum#184] [Bugs reports]
`<p><p><p><p><style>*{background:radial-gradient(1q,#6592CF,100px,#243D83)}p{padding:65;position:fixed;border:5vw solid#243D83;border-radius:50%;margin:-10;background:0;top:75;left:50}p+p{left:200}p+p+p{left:125;top:0}p+p+p+p{top:150`

that has 0 difference, yet get 99.9%
meanwhile gradients solutions flicker when hover over and gets 100% regardless, the tolerance of algorithm is more loose with gradients for no reason

**Comment by co0kie** [comment]
I'm also getting 99.9% using radial-gradient

**Reply by SelenIT** [reply]
Radial gradient with `69q` (approx. 65px) and `90q` (approx. 85px) as internal and external radii of the blue arcs works for me. But with `border`s or `box-shadow`s solutions with seemingly perfect match keep getting 99.9%, and neither of common workarounds (reducing `border-radius` from 50% to something like 49.8%, increasing/decreasing the shadow spread radius by 0.1px etc.) didn't help here...

**Comment by rafehimself** [comment]
I just tried improve score in #66 Batmicky and another 99.9% bug because I use border radius ...
sad that it is being ignored for such a long time ... 👀

**Reply by gunnarbittersmann** [reply]
Simple solution: if `border-radius` doesn’t work for you, don’t use `border-radius`.

Honestly, what do you want?
Make 99.9% the new 100%, i.e. full score with 99.9% matches? I have many 99.9% matches in the vaults that are much shorter than my current best 100% matches. But that’s the game: achieve 100%. And this shouldn‘t be changed.
Tweak the algorithm for each target, i.e. use a different algorithm for each target? No way, IMHO.

BTW, the algorithm _is_ very forgiving, especially when it comes to Batmicky. My solution has some visual differences and still matches to 100%. That’s part of the game: to find out where you have to be strict, and where you can be loose.

**Reply by SelenIT** [reply]
As I understand @rafehimself's complaint, it's the inconsistency of the result with the perception of the difference by the human eye in the first place. It's really annoying when the visually perfectly matching shape, with mathematically correct dimensions, gets rejected only because one pixel somewhere over there happened to be out of accepted range due to anti-aliasing algorithm, while the visibly mismatching shape (like my solution for target 92 where small arcs are not rounded at all) can be accepted as long as all the pixels fit into the ±1px acceptance interval. This just doesn't _feel_ right. Maybe the checking algorithm should somehow prioritize geometric precision over color perfectness (in all cases)?..

**Reply by rafehimself** [reply]
I have no idea what are trying to defending here?
are you the author of that flawed algorithm ? 🤣 

the code is made to server us, humans ..., not the other way around, as you try to convince it,
and if it doesn't provide results that are expected, it's a job of a developer to fix it
![image](https://user-images.githubusercontent.com/58576132/142856227-3719d960-1b36-4530-bde4-e8e6d4d95951.png)
as far as my reading comprehension skills goes
THE GOAL of GAME is to replicate the target as close as possible, 
and that should always match with 100% ,if there is close to 0 margin of error - in terms of visual effect - REGARDLESS of what method has been used.
the goal is not to crack the algorithm nor to study it in order to omit it, (which ppl do anyways due to human nature)
because that simply lead to absurd situations where an ideal replications are rejected and a tangible visual differences gets 100%

also that BUG is repetitive in multiple targets #41 #44 #50 #59 #66 #90 #92 #96, at least those are the one I'm sure of, other users might find even more

**Reply by verberg** [reply]
Have you seen the "top-solution" #91? If I load this code, it's not the "picture" like in the target  at all (sorry my english...). But if I submit this solution the match is 100%. I do not understand why this can happen!
Here this "top-sulution-code": <p>.<p>.<p>.<p>.<style>*{margin:75 0-65;background:radial-gradient(1q,#F6DF96 106q,#D25B70 0 775em,#6CB3A9)fixed;font:9%/5vh'

**Reply by rafehimself** [reply]
@verberg  that's the charm of using font based solutions, they will be render differently depending on the device default font

**Reply by SelenIT** [reply]
The current top solution for target 91 _should_ render quite close to the target image in up-to-date Chromium-based browsers. Actually, it doesn’t have anything font-specific in it (unlike the previous version that relied on the `ex` unit and minimum font setting). But apparently it exploits [a bug](https://github.com/cssbattle/forum/discussions/161#discussioncomment-1608645) in gradient rendering in Chrome, so doesn’t “work” in other browsers (which actually render it correctly according to the spec).

**Comment by Beowolve** [comment]
For me Batmicky was the last target where I reached 100%, I can feel your pain...

**Comment by verberg** [comment]
I have the same problem (not only with #96): I don't see any difference, it seems to be perfectly black and I get only 99.9%. This is very enoying:
Here ist my code (I also tryed others width box-shadows insted of borders, always  width 99.9%):
<p><p><style>*>*{position:fixed;background:radial-gradient(circle,#6592CF 100px,#243D83 65px)}p{background:#0000;width:130;height:130;border-radius:50%;border:20px solid#243D83;left:115;top:-26;-webkit-box-reflect:below -20px}p+p{top:49;left:40;-webkit-box-reflect:right -20px

**Reply by Beowolve** [reply]
-webkit-box-reflect has that problem a lot. My advice is to avoid it, too much characters anyway. Also for this target I suggest to you playing with radial-gradient offsets in shorthand notation.
Check out the other thread (https://github.com/cssbattle/forum/discussions/177), where I posted some examples about radial-gradient usage.

**Comment by himanshubhatt1** [comment]
How can i reduce this ....🤔


<style>*{background:radial-gradient(1q,#6592CF 0 100px,#243D83)}*>*{background:radial-gradient(1q,#0000,65px,#243D83 0 85px,#0000)-75px,radial-gradient(1q,#0000,65px,#243D83 0 85px,#0000)75px,radial-gradient(1q,#0000,65px,#243D83 0 85px,#0000)0 75px,radial-gradient(1q,#0000,65px,#243D83 0 85px,#0000)0 -75px

**Reply by Beowolve** [reply]
You are basically on the right track. The micro optimizations, like changing the `px` to `q` will save you a few chars, but it's also possible to reduce the number of `radial-gradients` in your code. In shorthand notation, there is more than just offsets. You can also change the size of the viewport you are drawing at with `/<width><height>`. This is not radial-gradient specific. It's a background feature, omitting background-size for shorter code.

**Reply by rafehimself** [reply]
the biggest change would be reducing amounts of gradients being used

**Comment by gunnarbittersmann** [comment]
Like every test the scoring algorithm comes with false positives and false negatives.

False positive means: visible difference, but 100% match – which we happily take advantage of. *(“Hey Batmicky, you have such beautiful ears.”)* 😏 Many records come from tweaking values (like 99 for 100, or 21q for 20px), watching that you still get 100%.

False negative means: no visible difference, but only 99.9% match (or less) – that’s the annoying part.

Changing the algorithm would have consequences. Should solutions that have matched to 99.9% before become 100% matches, many players who have rejected a way of solving this target because they ran into the 99.9% trap would not be aware that this way would now lead to a 100% solution. This would be unfair.

We have been there before: With an update of the Chromium version, suddenly the solution to [target #1](https://cssbattle.dev/play/1) could be made 2 characters shorter replacing `img` with `i` (or any other  except `p` and `q`). After another update the shortest solution to [open target #62](https://cssbattle.dev/play/62) doesn’t work anymore.

That were just Chromium updates. Changing the algorithm would be much worse.

**Reply by SelenIT** [reply]
Actually, we had algorithm change too, in the very first days of CSSBattle, with only 12 original targets available and SVGs not banned yet. With the initial release, Target 10 (and I guess some others) were virtually "undoable". @cssbattle team made a special announce for the players who had "almost working" solutions with 99.9% match that they can resubmit them and give them a second chance. TBH, the current situation with border-based solutions for target 96 (as well as target 44 and some others) sometimes reminds me of the situation prior to that change.

IMHO, it's OK to have _some_ false negatives. The question is, which share of false negatives (as well as false positives) should be acceptable... and how to measure it 😳🤔

**Reply by gunnarbittersmann** [reply]
Acceptable – for whom?

People play the targets for different reasons. I’d say there are two major groups:

1. People who just want to solve the targets somehow. Multi-line code with `div` (as if there weren’t other HTML element types) and `class` (as if there weren’t other options to select). There are some videos on YouTube – and [I’m not in their target audience.](https://twitter.com/g16n/status/1455496339568177158)
Of course a 100% result would give them the most satisfaction. But I think they can live with 99.9% as well when the diff in the preview is completely black.

2. The code-golfers aiming at the shortest code possible. Because of the scoring (up to 99.9% matches – less than 600 points; 100% matches – more than 600 points) any result under 100% is unacceptable for them.
They try everything to trick the algorithm. So it’s OK when it goes both ways: sometime they get tricked by the algorithm. This adds another level to the challenge: if one way doesn’t lead to a 100% match, find another way. Or tweak the values – [target #29](https://cssbattle.dev/play/29) comes to mind.
I think for this group it’s totally acceptable when the first approach doesn’t lead to the goal.

**Reply by rafehimself** [reply]
Changes are inevitable, there is no future where chromium or css will stop updating/evolving, current version is already outdated because outline doesn't work as it should and new features like rotate:45deg are on the horizon

holding to old ways will only make it a dinosaur, and we all know how they end up

as long as new solutions that used be 99.9% despite of visual match will be available for everyone it's totally fair to fix that,
just treat it like your favorite motto a new challenge😉 
what is unfair it's a situation when some old scores based on svg or code like #61 it's not available anymore for new comers to make, but they are still kept in a leaderboard, as well as abuse of svg path behind the scene with a trickery, which is discuss here https://github.com/cssbattle/forum/discussions/148

less 99.9% would provide a better experience for everyone
and absurd situations like that the algorithm accept border as 100% but not a background with that color and the same pixel values less annoying
